File: Daily Challanges/2026-07-July/2026-07-05.js
function bucketFill(grid, start, newValue) {
  const rows = grid.length;
  const cols = grid[0].length;
  const [row, col] = start;
  const oldValue = grid[row][col];

  if (oldValue === newValue) {
    return grid;
  }

  const stack = [[row, col]];

  while (stack.length) {
    const [r, c] = stack.pop();

    if (r < 0 || r >= rows || c < 0 || c >= cols) {
      continue;
    }

    if (grid[r][c] !== oldValue) {
      continue;
    }

    grid[r][c] = newValue;

    stack.push([r + 1, c]);
    stack.push([r - 1, c]);
    stack.push([r, c + 1]);
    stack.push([r, c - 1]);
  }

  return grid;
}

console.log(
  bucketFill(
    [
      ["R", "G"],
      ["R", "G"],
    ],
    [0, 1],
    "B",
  ),
);
console.log(
  bucketFill(
    [
      ["Y", "G", "G"],
      ["Y", "Y", "Y"],
      ["B", "Y", "R"],
    ],
    [1, 2],
    "B",
  ),
);
console.log(
  bucketFill(
    [
      ["O", "O", "P"],
      ["P", "O", "O"],
      ["P", "P", "O"],
    ],
    [2, 0],
    "R",
  ),
);
console.log(
  bucketFill(
    [
      ["T", "T", "R", "T"],
      ["R", "T", "R", "T"],
      ["R", "T", "R", "T"],
      ["T", "T", "T", "T"],
    ],
    [0, 3],
    "Y",
  ),
);
console.log(
  bucketFill(
    [
      ["G", "B", "G", "B"],
      ["R", "B", "B", "G"],
      ["B", "G", "B", "R"],
      ["B", "G", "G", "B"],
    ],
    [2, 2],
    "G",
  ),
);

File: Daily Challanges/2026-07-July/2026-07-05.py
def bucket_fill(grid, start, new_value):
    rows = len(grid)
    cols = len(grid[0])
    row, col = start
    old_value = grid[row][col]

    if old_value == new_value:
        return grid

    stack = [(row, col)]

    while stack:
        r, c = stack.pop()

        if r < 0 or r >= rows or c < 0 or c >= cols:
            continue

        if grid[r][c] != old_value:
            continue

        grid[r][c] = new_value

        stack.append((r + 1, c))
        stack.append((r - 1, c))
        stack.append((r, c + 1))
        stack.append((r, c - 1))

    return grid


print(bucket_fill([["R", "G"], ["R", "G"]], [0, 1], "B"))
print(bucket_fill([["Y", "G", "G"], ["Y", "Y", "Y"], ["B", "Y", "R"]], [1, 2], "B"))
print(bucket_fill([["O", "O", "P"], ["P", "O", "O"], ["P", "P", "O"]], [2, 0], "R"))
print(
    bucket_fill(
        [
            ["T", "T", "R", "T"],
            ["R", "T", "R", "T"],
            ["R", "T", "R", "T"],
            ["T", "T", "T", "T"],
        ],
        [0, 3],
        "Y",
    )
)
print(
    bucket_fill(
        [
            ["G", "B", "G", "B"],
            ["R", "B", "B", "G"],
            ["B", "G", "B", "R"],
            ["B", "G", "G", "B"],
        ],
        [2, 2],
        "G",
    )
)
